File: Java_Advanced/gsapi/src/main/java/br/com/fiap/gs/gsapi/dto/request/EnderecoRequestDTO.java
package br.com.fiap.gs.gsapi.dto.request;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Size;
import jakarta.validation.constraints.Digits;


public class EnderecoRequestDTO {

    @NotBlank(message = "O CEP não pode estar em branco.")
    @Pattern(regexp = "^\\d{5}-?\\d{3}$", message = "O CEP deve estar no formato XXXXX-XXX ou XXXXXXXX.")
    private String cep;

    @NotNull(message = "O número não pode ser nulo.")
    // @Digits(integer = 5, fraction = 0, message = "O número deve ser um inteiro de até 5 dígitos.") // Se fosse string
    private Integer numero; // NUMBER(5)

    @NotBlank(message = "O logradouro não pode estar em branco.")
    @Size(max = 255, message = "O logradouro não pode exceder 255 caracteres.")
    private String logradouro;

    @NotBlank(message = "O bairro não pode estar em branco.")
    @Size(max = 255, message = "O bairro não pode exceder 255 caracteres.")
    private String bairro;

    @NotBlank(message = "A localidade (cidade) não pode estar em branco.")
    @Size(max = 100, message = "A localidade não pode exceder 100 caracteres.")
    private String localidade;

    @NotBlank(message = "A UF não pode estar em branco.")
    @Size(min = 2, max = 2, message = "A UF deve ter 2 caracteres.")
    private String uf;

    @NotBlank(message = "O complemento não pode estar em branco.") // DDL diz NOT NULL
    @Size(max = 255, message = "O complemento não pode exceder 255 caracteres.")
    private String complemento;

    // Latitude e Longitude serão preenchidas pelo ViaCEP/outra API,
    // então não são obrigatórias na requisição inicial do usuário,
    // mas podem ser enviadas se já conhecidas.
    @NotNull(message = "Latitude não pode ser nula.")
    @Digits(integer = 3, fraction = 7, message = "Latitude inválida.") // ex: -90.1234567
    private Double latitude;

    @NotNull(message = "Longitude não pode ser nula.")
    @Digits(integer = 4, fraction = 7, message = "Longitude inválida.") // ex: -180.1234567
    private Double longitude;


    // Getters e Setters
    public String getCep() { return cep; }
    public void setCep(String cep) { this.cep = cep; }
    public Integer getNumero() { return numero; }
    public void setNumero(Integer numero) { this.numero = numero; }
    public String getLogradouro() { return logradouro; }
    public void setLogradouro(String logradouro) { this.logradouro = logradouro; }
    public String getBairro() { return bairro; }
    public void setBairro(String bairro) { this.bairro = bairro; }
    public String getLocalidade() { return localidade; }
    public void setLocalidade(String localidade) { this.localidade = localidade; }
    public String getUf() { return uf; }
    public void setUf(String uf) { this.uf = uf; }
    public String getComplemento() { return complemento; }
    public void setComplemento(String complemento) { this.complemento = complemento; }
    public Double getLatitude() { return latitude; }
    public void setLatitude(Double latitude) { this.latitude = latitude; }
    public Double getLongitude() { return longitude; }
    public void setLongitude(Double longitude) { this.longitude = longitude; }
}

File: Advanced_Business_Development_with.NET/gsApi/gsApi/dto/request/EnderecoRequestDto.cs
// File: gsApi/dto/request/EnderecoRequestDto.cs
using System.ComponentModel.DataAnnotations;

namespace gsApi.DTOs.Request
{
    public class EnderecoRequestDto
    {
        [Required(ErrorMessage = "O CEP não pode estar em branco.")]
        [RegularExpression(@"^\d{5}-?\d{3}$", ErrorMessage = "O CEP deve estar no formato XXXXX-XXX ou XXXXXXXX.")]
        public required string Cep { get; set; }

        [Required(ErrorMessage = "O número não pode ser nulo.")]
        [Range(1, 99999, ErrorMessage = "O número deve ser um inteiro de até 5 dígitos.")]
        public int Numero { get; set; } // Tipo de valor, 'required' C# não se aplica para nulidade

        [Required(ErrorMessage = "O logradouro não pode estar em branco.")]
        [StringLength(255, ErrorMessage = "O logradouro não pode exceder 255 caracteres.")]
        public required string Logradouro { get; set; }

        [Required(ErrorMessage = "O bairro não pode estar em branco.")]
        [StringLength(255, ErrorMessage = "O bairro não pode exceder 255 caracteres.")]
        public required string Bairro { get; set; }

        [Required(ErrorMessage = "A localidade (cidade) não pode estar em branco.")]
        [StringLength(100, ErrorMessage = "A localidade não pode exceder 100 caracteres.")]
        public required string Localidade { get; set; }

        [Required(ErrorMessage = "A UF não pode estar em branco.")]
        [StringLength(2, MinimumLength = 2, ErrorMessage = "A UF deve ter 2 caracteres.")]
        public required string Uf { get; set; }

        [Required(ErrorMessage = "O complemento não pode estar em branco.")]
        [StringLength(255, ErrorMessage = "O complemento não pode exceder 255 caracteres.")]
        public required string Complemento { get; set; }

        [Required(ErrorMessage = "Latitude não pode ser nula.")]
        [Range(-90.0, 90.0, ErrorMessage = "Latitude inválida. Deve estar entre -90 e 90.")]
        public double Latitude { get; set; } // Tipo de valor

        [Required(ErrorMessage = "Longitude não pode ser nula.")]
        [Range(-180.0, 180.0, ErrorMessage = "Longitude inválida. Deve estar entre -180 e 180.")]
        public double Longitude { get; set; } // Tipo de valor
    }
}